documentation/src/main/asciidoc/chapter-10-advanced-mapping-options.asciidoc

```diff
@@ -110,6 +110,40 @@ public interface SourceTargetMapper {
 
 The example demonstrates how to use defaultExpression to set an `ID` field if the source field is null, this could be used to take the existing `sourceId` from the source object if it is set, or create a new `Id` if it isn't. Please note that the fully qualified package name is specified because MapStruct does not take care of the import of the `UUID` class (unless it’s used otherwise explicitly in the `SourceTargetMapper`). This can be resolved by defining imports on the @Mapper annotation (see <<expressions>>).
 
+[[sub-class-mappings]]
+=== Subclass Mapping
+
+When both input and result types have an inheritance relation, you would want the correct specialization be mapped to the matching specialization. Suppose an `Apple` and a `Banana`, which are both specializations of `Fruit`.
+
+.Specifying the sub class mappings of a fruit mapping
+====
+[source, java, linenums]
+[subs="verbatim,attributes"]
+----
+@Mapper
+public interface FruitMapper {
+
+    @SubClassMapping( sourceClass = AppleDto.class, targetClass = Apple.class )
+    @SubClassMapping( sourceClass = BananaDto.class, targetClass = Banana.class )
+    Fruit map( FruitDto source );
+
+}
+----
+====
+
+If you would just use a normal mapping both the `AppleDto` and the `BananaDto` would be made into a `Fruit` object, instead of an `Apple` and a `Banana` object. By using the subclass mapping an `AppleDtoToApple` mapping will be used for `AppleDto` objects, and an `BananaDtoToBanana` mapping will be used for `BananaDto` objects. If you would try to map a `GrapeDto` it would still turn into a `Fruit`. In the case that the `Fruit` is an abstract class or an interface, you would get an `IllegalArgumentException` because it is unknown how to map a `GrapeDto`. Adding the missing (`@SubClassMapping`) for it will fix that.
+
+[TIP]
+====
+If the mapping method for the subClasses does not exist it will be created and any other annotations on the fruit mapping method are inherited by the newly generated mappings.
+====
+
+[NOTE]
+====
+Combining `@SubClassMapping` with update methods is not supported. If you use an update method with a return value then the subclass mappings will return a new instance based on the mapping, but it will not take into account the `@MappingTarget` parameter.
+The same issue exists for the `@Context` and `@TargetType` parameters.
```

Comment:
But then you would also need to check if the `@MappingTarget` is of the same type as the `@SubClassMapping(target=TargetSub.class, source=SourceSub.class)` so you would end up with 2 instance of checks in order to call the subclass method which would then become: `void updateTargetSubWithSourceSub(@MappingTarget TargetSub target, SourceSub source)`. So the implementation of the mapping method calling that update method would look something like this:
```java
void updateMapping( @mappingTarget Target target, Source source ) {
    if ( source instanceof SourceSub && target instanceof TargetSub ) {
        updateTargetSubWithSourceSub( (TargetSub) target, (SourceSub) source );
    } else if ( source instanceof SourceSub) {
        // what now??? source argument matches subclass of sourceClass but target argument doesn't match subclass of targetClass.
    } else {
        // default update behavior
    }
}
```
But how to search for an existing method or method definition that can be called?

I think it would indeed be best to have that as another Issue: support for subclassmappings in combination with update mappings. Should I create an issue for this so that we can discuss this topic in that issue's thread?